Do not load whole jar file when --function is given

antlechner opened the pull request:
Only start the analysis of loading classes at the main class, i.e. the class that contains the entry method.

We currently load a lot of classes that we don't need when jbmc is given a jar file as an argument. This change prevents that, and running on a .jar should get performance close to running on a .class.

I don't think we have any tests for large jar files. Locally I got very promising results with these changes, and I will try running them on some company benchmarks. If the results from those runs look good, I will remove the "do not merge" label and suggest we merge it, and worry later about how to write good tests for large jar files (but we should definitely do that too).

antlechner committed:
Do not load whole jar file when --function is given

Only start the analysis of loading classes at the main class, i.e. the
class that contains the entry method.

mgudemann commented on jbmc/src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -213,6 +213,13 @@ bool java_bytecode_languaget::parse(
     {
+      const std::string &entry_method = config.main;
+      // If we have an entry method, we can derive a main class.
+      if(entry_method != "")
```
use `!entry_method.empty()` here

antlechner replied:
Done

mgudemann commented on jbmc/src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -213,6 +213,13 @@ bool java_bytecode_languaget::parse(
+      {
+        const auto last_dot_position = entry_method.find_last_of('.');
+        main_class = entry_method.substr(0, last_dot_position);
+      }
```
maybe reading the manifest should be in an `else` branch. As it is currently, a main entry in the manifest would override the setting from `--function`

antlechner replied:
Done

smowton reviewed (commented):
Add a regression test, using `--verbosity 10` so you can observe the "loading class from JAR" lines -- you can for example check that a class present in a jar but not referenced from the entry point doesn't get loaded.

mgudemann reviewed (approved):
LGTM

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: f4f9ca0b4487e404bcbe3b73db94b7f65c02be4a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84990112

antlechner committed:
Test class loading for jar file arguments

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 06481d7814c4aefb23b05a37c6ed4ebb47b92b52).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85051606

allredj reviewed (commented):
Would it be useful to document that behaviour somewhere in the CProver manual?

antlechner committed:
Document the new class loading strategy for JARs

antlechner commented:
@smowton I added a test as you suggested.
@allredj I added some doxygen above `java_bytecode_languaget::parse` and also updated the output of `jbmc --help` to explain that jbmc can run on .class and .jar files. @peterschrammel said that the output `--janalyzer --help` should be updated accordingly, so I made sure the two are the same.
Running this on benchmarks looks promising (not a huge improvement but it definitely hasn't gotten worse), so this is ready for re-review now.

peterschrammel commented on jbmc/src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -207,12 +207,22 @@ bool java_bytecode_languaget::parse(
-      if(manifest_main_class!="")
-        main_class=manifest_main_class;
+        // if the manifest declares a Main-Class line, we got a main class
+        if(manifest_main_class != "")
```
! .empty()

antlechner replied:
Done

allredj commented on jbmc/src/janalyzer/janalyzer_parse_options.cpp:
```diff
@@ -733,7 +733,14 @@ void janalyzer_parse_optionst::help()
+    " manifest (checked in\n"
+    "                              this order), the behavior is the same as"
+    " running janalyzer\n"
+    "                              on the corresponding class file."
```
:pick: line wrapping is weird

allredj replied:
Does clang let you indent the (for instance) `" running janalyzer` line?

antlechner replied:
clang isn't a problem, CLion is doing the weird line wrapping thing. What was already in the file doesn't look much nicer but I can try to format the new lines in the same way using vim.

allredj commented on jbmc/src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -1022,7 +1022,14 @@ void jbmc_parse_optionst::help()
+    "                              from --main-class, --function or the JAR"
+    " manifest (checked in\n"
+    "                              this order), the behavior is the same as"
+    " running jbmc on the\n"
```
:pick: line wrapping is weird

antlechner replied:
Fixed (I think - it's similar to what was done below in the same file).

peterschrammel commented on jbmc/src/janalyzer/janalyzer_parse_options.cpp:
```diff
@@ -733,7 +733,14 @@ void janalyzer_parse_optionst::help()
-    " janalyzer class              name of class to be checked\n"
+    " janalyzer class              name of class or JAR file to be checked\n"
+    "                              In the case of a JAR file, if a main class"
+    " can be inferred\n"
```
The help text should be broken after 80 chars and properly aligned.

antlechner replied:
I think the formatting is correct now (as it was done below in the same file).

allredj commented on jbmc/src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -1022,7 +1022,14 @@ void jbmc_parse_optionst::help()
+    " jbmc class                   name of class or JAR file to be checked\n"
+    "                              In the case of a JAR file, if a main class"
+    " can be inferred\n"
+    "                              from --main-class, --function or the JAR"
```
:pick: Oxford comma after `--function` helps understanding I think.

antlechner replied:
Done

allredj reviewed (commented):
Do you have a test-gen update?

antlechner commented:
@allredj Yes, it's very old and you approved it a long time ago already. :)
https://github.com/diffblue/test-gen/pull/2181

allredj reviewed (approved):
Looks good!

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 3c9b025c265b90bae5cb9e00e4f69bde50bbeb61).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85303843
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 117457d4de8ec2972c8269f9636176cff8fb31a8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85320651

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 7d7424f081f65d99249a4989b1b781d75c21acec).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85559729
